<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This package reads whole genome multilocus sequence
    typing profiles with food sources to create a random
    forest model. Then, another wgMLST profile can query the
    model to yield a prediction of which food source it might have
    come from.">
<title>Predicts Food Source from an MLST Profile • sourcerer</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Predicts Food Source from an MLST Profile">
<meta property="og:description" content="This package reads whole genome multilocus sequence
    typing profiles with food sources to create a random
    forest model. Then, another wgMLST profile can query the
    model to yield a prediction of which food source it might have
    come from.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">sourcerer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ncezid-biome/source-attribution-MLST/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="source-attribution-mlst">source-attribution-MLST<a class="anchor" aria-label="anchor" href="#source-attribution-mlst"></a>
</h1></div>
<p>This package implements the ideas from the source attributions project in <a href="https://doi.org/10.1089/fpd.2023.0046" class="external-link">Gu et al 2023</a>. Given a set of known bacterial foodborne pathogens and their MLST profile, this package can predict the food sources of future bacterial foodborne pathogen genomes.</p>
<ol style="list-style-type: decimal">
<li>Create a set of random forest models</li>
<li>Query your genome against the random forest models</li>
<li>Aggregate the predictions</li>
</ol>
<pre class="mermaid"><code>flowchart LR
    A[MLST profiles of bacterial foodborne pathogens with known etiology] 
        --&gt; |"bootstrap()"| B{{random forest models}}  
    C[MLST profile of new bacterial foodborn pathogen]
        --&gt; |"predict()"| B
    B --&gt; |"predict()"| D{{Predictions}}
    D --&gt; |"aggregate_predictions()"| E{{Aggregate stats:\n* Best estimates of food vehicle\n* Rankings of most determinant loci}}
    B --&gt; |"aggregate_models()" | F{{Aggregate reference model\n* Most important genes}}</code></pre>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Requires R and <code>devtools</code></p>
<pre class="shell"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ncezid-biome/source-attribution-MLST"</span><span class="op">)</span></span></code></pre>
<p>To install the executable scripts, locate where they were installed, e.g., <code>$HOME/R/4/3.2/sourcerer</code>, and update your <code>PATH</code>. In Linux, it looks like this:</p>
<pre class="shell"><code># Find the installation
find ~ -type d -name sourcerer
# =&gt; assume the path is $HOME/R/4/3.2/sourcerer for this example
export PATH=$PATH:$HOME/R/4/3.2/sourcerer/exec
# Double check the installation with the which command
which bootstrapRF.R
which predictRF.R
</code></pre>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>First, create a model or models with <code><a href="reference/bootstrapping.html">bootstrapping()</a></code> or <code>Rscript bootstrapRF.R</code>. Then, query the model with <code><a href="reference/prediction.html">prediction()</a></code> or <code>Rscript predictRF.R</code>.</p>
<p>Creating a model is done with a “bootstrap” because it represents only one outcome out of all the stochatic outcomes in the random forest model. Therefore, the usage is to create many models and then run prediction on many models to get an aggregated outcome.</p>
<div class="section level3">
<h3 id="examples-with-api">Examples with API<a class="anchor" aria-label="anchor" href="#examples-with-api"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ncezid-biome.github.io/source-attribution-MLST">"sourcerer"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">reps</span>    <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">my_seed</span> <span class="op">&lt;-</span> <span class="fl">23</span> <span class="co"># leave blank for random seed</span></span>
<span><span class="va">ncores</span>  <span class="op">&lt;-</span> <span class="fl">1</span>  <span class="co"># parallelization</span></span>
<span></span>
<span><span class="co"># Random forest modeling step</span></span>
<span><span class="va">rf_filenames</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/bootstrapping.html">bootstrapping</a></span><span class="op">(</span></span>
<span>                   input  <span class="op">=</span> <span class="st">"tests/testthat/Listeria_isolates.csv.gz"</span>,</span>
<span>                   output <span class="op">=</span> <span class="st">"results/"</span>,</span>
<span>                   ncores <span class="op">=</span> <span class="va">ncores</span>,</span>
<span>                   bootstrap_reps  <span class="op">=</span> <span class="va">reps</span>,</span>
<span>                   loci_start_with <span class="op">=</span> <span class="st">"LMO"</span>,</span>
<span>                   my_seed <span class="op">=</span> <span class="va">my_seed</span></span>
<span>                <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Make predictions, one per model</span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">rf_filenames</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">predictions</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/prediction.html">prediction</a></span><span class="op">(</span></span>
<span>        model_filename <span class="op">=</span> <span class="va">rf_filenames</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,</span>
<span>        query <span class="op">=</span> <span class="st">"tests/testthat/example_query.csv"</span>, </span>
<span>        ncores <span class="op">=</span> <span class="va">ncores</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Get a table of predictions</span></span>
<span><span class="va">aggregate_predictions_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/aggregate_predictions.html">aggregate_predictions</a></span><span class="op">(</span></span>
<span>                                  predictions <span class="op">=</span> <span class="va">predictions</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">aggregate_predictions_table</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Aggregate the models to find the most important genes</span></span>
<span><span class="va">ref_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/aggregate_model.html">aggregate_model</a></span><span class="op">(</span><span class="va">rf_filenames</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ref_model</span><span class="op">$</span><span class="va">aggregate_rank</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/ncezid-biome/source-attribution-MLST/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ncezid-biome/source-attribution-MLST/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sourcerer</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lee Katz <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2533-9161" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Zhaohui Cui <br><small class="roles"> Author </small>  </li>
<li>Sean Browning <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-4580-1843" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lee Katz, Zhaohui Cui, Sean Browning.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
